### ISSUE TITLE
Add converters betwee `FloatingPointValue` and `FixedPointValue`

### ISSUE THREAD

**Issue by mkorbel1** [intel/rohd-hcl#112]
### Motivation

Not all numbers can fit in `double`, so converting back and forth from `double` doesn't cover all cases

### Desired solution

Add automation to convert between `FloatingPointValue` and `FixedPointValue`

### Additional details

https://github.com/intel/rohd-hcl/pull/99

**Comment by mkorbel1** [comment]
Related issues from the hardware side: #116, #123

**Comment by sweeksBigBlue** [comment]
Where exactly are you thinking of adding this logic? You want to go both directions as in Floating Point <--> Fixed Point? Are these only to be IEEE standard or do we want the conversion to be between two numbers based on sort of values provided?

**Comment by mkorbel1** [comment]
I guess these could be just functions on the types directly?  Like within `FloatingPointValue`, a `toFixedPointValue`, and vice versa?

**Comment by mkorbel1** [comment]
There probably need to be some arguments regarding rounding modes and precisions/widths

**Comment by jcfarwe** [comment]
@sweeksBigBlue and I will claim this issue and work on it.

**Comment by mkorbel1** [comment]
great!  thank you!

**Comment by sweeksBigBlue** [comment]
Starting our work on this but it appears @desmonddak has done the fixed to float, so are we just going to focus on floating point to fixed for our solution @mkorbel1?

**Comment by desmonddak** [comment]
This is the entire attempt by me and it is bug-riddled.  You could grab this in your own PR and free me from my addled attempt.

```dart
floating_point_value.dart: 
 /// Losslessly convert a [FloatingPointValue] to a [FixedPointValue].
  FixedPointValue toFixedPointValue() {
    // space for full shift (bias*2), mantissa, sign and jbit
    var fxdMantissa = [
      if (isNormal()) LogicValue.one else LogicValue.zero,
      mantissa
    ].swizzle().zeroExtend(bias * 2 + mantissaWidth + 2);
    fxdMantissa = sign == LogicValue.one
        ? ~fxdMantissa + 1
        : fxdMantissa; // make sure the sign is correct
    final shift = exponent.toInt() - 1;
    final shiftedFxdMantissa =
        exponent.toInt() > 0 ? fxdMantissa << shift : fxdMantissa;

    final fxpN = mantissaWidth + 2;
    final fxpM = fxdMantissa.width - fxpN - 1;
    return FixedPointValue(
        value: shiftedFxdMantissa, m: fxpM, n: fxpN, signed: true);
  }

**Comment by sweeksBigBlue** [comment]
Okay, I think I'm following but I'm going to verify some things here to make sure I don't take us off on the wrong route. So these functions already exist and you need help getting them to the correct value? You do want to handle rounding mode for the fixed to float values? (We were just playing around with that and I was going really simple)? I'm sure we will have more questions as we bite this off but just want to check.

Adding @curtisan-intel for visibility as he usually asks better questions than I do.

**Comment by desmonddak** [comment]
No, these methods don't exist except for the attempt above (FloatingPointValue.toFixedPointValue).  This one should be lossless (no rounding).  Very wide output.  It is buggy.  Two tests exist in test/arithmetic/values/floating_point_value.test at the very end.  Fiddle with the mantissa width and you will get bugs.

```dart
test('FloatingPointValue to FixedPointValue roundtrip exhaustive'
```

The other methods needed use the Populator pattern, e.g., as in floating_point_value_populator.dart.  They would need to use a rounding component -- we need one modularized (if you have time) out of existing code:  LogicValue input, mode, lsbPos are the inputs.
```
FloatingPointValue.ofFixedPointValue(FixedPointValue fxv);
FixedPointValue.ofFloatingPointValue(FloatingPointValue fpv);
```

So along the way, if you are interested, we need:
```dart
bool roundPredicate (LogicValue inp,  FloatingPointRoundingMode mode,  int mantissaWidth, int lsbPos);
```
This can just support RNE and call some code that is in ofDouble():  Basically look for LGRS code that does LSB/Guard/Round/Sticky) style rounding determination.

Does this help?

**Comment by desmonddak** [comment]
I guess I am now referring to a second issue:  https://github.com/intel/rohd-hcl/issues/191

This is modularizing rounding.  It can happen on both the value side and on the component side.  
If you are working on the value side (the converter above), the rounding modularization can happen as well as part of that (or at least  construction of a modular value-side rounding component to get us going).  So this would be in addition a nice-to-have that is in our issues list.

**Comment by sweeksBigBlue** [comment]
Okay, I think I'm getting it! I might break this issue up into smaller multiple PRs as we work on it so we can get thorough discussions on how this should work and if it fits the schemes you are imagining. Right now working on the fixed to float component for the value. Which brings up one thing. our find first one looks like it only goes from LSB forward and it would be nice if it went in reverse, so I'm debating messing around with that to make our lives easier.

**Comment by jcfarwe** [comment]
Slightly higher level, do we want to support converting a Fixed to a Float of a known format (8_e4m3, 8_e5m2, bf16, etc), or arbitrary formats, or both?
